src/Init/Lean/PrettyPrinter/Parenthesizer.lean

```diff
@@ -0,0 +1,444 @@
+  visit (mkConst stx.getKind);
+⟨_, some lbp', trailRbp'⟩ ← get
+  | panic! "visitParenthesizable: visited a term without tokens?!";
+trace! `PrettyPrinter.parenthesize ("...precedences are " ++ fmt rbp ++ " >? " ++ fmt lbp' ++ ", " ++ fmt trailRbp' ++ " <? " ++ fmt lbp);
+-- Should we parenthesize?
+trailRbp' ← if rbp > lbp' || (match trailRbp', lbp with some trailRbp', some lbp => trailRbp' < lbp | _, _ => false) then do
+    -- The recursive `visit` call, by the invariant, has moved to the next node to the left. In order to parenthesize
+    -- the original node, we must first move to the right, except if we already were at the left-most child in the first
+    -- place.
+    when (idx > 0) goRight;
+    setCur (mkParen stx);
+    goLeft;
+    -- after parenthesization, there is no more trailing parser
+    pure none
+  else pure trailRbp';
+-- If we already had a token at this level (`lbp ≠ none`), keep the trailing parser. Otherwise, use the minimum of
+-- `rbp` and `trailRbp'`.
+let trailRbp := match trailRbp', lbp with
+  | _,              some _ => trailRbp
+  | some trailRbp', _      => some (Nat.min trailRbp' rbp)
+  | _,              _      => some rbp;
+modify (fun st => { st with lbp := lbp' <|> lbp, trailRbp := trailRbp })
+
+/-- Set token precedence and advance to the left. -/
+def visitToken (lbp : Nat) : ParenthesizerM Unit := do
+modify (fun st => { st with lbp := lbp });
+goLeft
+
+/-- Execute `x` at the right-most child of the current node, if any, then advance to the left. -/
+def visitArgs (x : ParenthesizerM Unit) : ParenthesizerM Unit := do
+stx ← getCur;
+when (stx.getArgs.size > 0) $
+  goDown (stx.getArgs.size - 1) *> x <* goUp;
+goLeft
+
+def evalNat (e : Expr) : ParenthesizerM Nat := do
+e ← liftM $ whnf e;
+some n ← pure $ Meta.evalNat e
+  | throw $ Exception.other $ "failed to evaluate Nat argument: " ++ toString e;
+pure n
+
+def evalOptPrec (e : Expr) : ParenthesizerM Nat := do
+e ← liftM $ whnf e;
+match e.getAppFn.constName? with
+| some `Option.none => pure 0
+| some `Option.some => evalNat e.appArg!
+| _ => throw $ Exception.other $ "failed to evaluate precedence: " ++ toString e
+
+def evalString (e : Expr) : ParenthesizerM String := do
+Expr.lit (Literal.strVal s) _ ← liftM $ whnf e
+  | throw $ Exception.other $ "failed to evaluate String argument: " ++ toString e;
+pure s
+
+partial def evalName : Expr → ParenthesizerM Name | e => do
+e ← liftM $ whnf e;
+if e.isAppOfArity `Lean.Name.anonymous 0 then
+  pure Name.anonymous
+else if e.isAppOfArity `Lean.Name.str 3 then do
+  n ← evalName $ e.getArg! 0;
+  s ← evalString $ e.getArg! 1;
+  pure $ mkNameStr n s
+else if e.isAppOfArity `Lean.Name.num 3 then do
+  n ← evalName $ e.getArg! 0;
+  u ← evalNat $ e.getArg! 1;
+  pure $ mkNameNum n u
+else
+  throw $ Exception.other $ "failed to evaluate Name argument: " ++ toString e
+
+@[builtinParenthesizer termParser]
+def termParser.parenthesizer : Parenthesizer | p => do
```

Comment:
The `... | p => do` is a nonstandard coding style :) We don't use it in the rest of the codebase. We use
```lean
def termParser.parenthesizer : Parenthesizer 
| p => do
...
```

Reply:
Oops, I may have picked that up from @gebner (or the other way around), since now it has also leaked into the codebase via @mhuisi's port of Gabriel's code :) .
![image](https://user-images.githubusercontent.com/109126/81066886-1ab25480-8ede-11ea-94d4-0711e901efe0.png)
It *is* conveniently compact